plugins/pulp_rpm/plugins/db/models.py

```diff
@@ -236,55 +461,114 @@ def rpm_search_dicts(self):
 
 
 class PackageGroup(Package):
-    UNIT_KEY_NAMES = ('id', 'repo_id')
-    TYPE = ids.TYPE_ID_PKG_GROUP
-
-    def __init__(self, id, repo_id, metadata):
-        Package.__init__(self, locals())
-        # these attributes should default to False based on yum.comps.Group.parse
-        for name in ('default', 'user_visible'):
-            if self.metadata.get(name) is None:
-                self.metadata[name] = False
+    # TODO add docstring to this class
+    package_group_id = mongoengine.StringField(required=True)
+    repo_id = mongoengine.StringField(required=True)
```

Comment:
Sit down for this one. ;) This data model also has existed for longer than I've been on the team, but in this case I at least know the explanation.

It's very difficult to come up with a decent unit key for a group that is globally unique. The best anyone could do is guarantee that the package_group_id will be unique within any particular repo.

So... the solution was to include the repo_id in unit key. And whenever a group gets copied to another repo in an "associate" operation, the importer actually makes a new Unit with exactly the same attributes, but changes the repo_id.

What I really wish we'd done is treated the entire comps.xml file as a unit, and not cared about what's in it. Maybe in pulp 3...

TL;DR it's a hack to facilitate global uniqueness.